service/tablet_allocator.cc

```diff
@@ -1009,12 +1009,46 @@ class load_balancer {
                     // FIXME: we can try another pair of non-colocated replicas of same sibling tablets.
                     lblogger.debug("Load limit reached, unable to emit migration for replica ({}, {}) to co-habit the replica ({}, {})",
                         t2_id, src, t1_id, dst);
-                    return;
+                    return false;
                 }
                 apply_load(nodes, mig_streaming_info);
 
+                utils::get_local_injector().inject("disallow_cross_rack_colocation", [&] {
+                    auto& topology = _tm->get_topology();
+                    if (topology.get_rack(src.host) != topology.get_rack(dst.host)) {
+                        on_fatal_internal_error(lblogger, "Cross rack colocation is not allowed, killing the node");
+                    }
+                });
+
                 lblogger.info("Created migration for replica ({}, {}) to co-habit same shard as ({}, {})", t2_id, src, t1_id, dst);
                 plan.add(std::move(mig));
+                return true;
+            };
+
+            // Enable rack aware colocation if RF is a multiple of number of racks,
+            // meaning that sibling tablets will have the same number of replicas
+            // on each rack, allowing co-location migrations to not cross rack
+            // boundaries.
+            auto rack_aware_colocation = [&] () {
+                auto& topology = _tm->get_topology();
+                auto [_, rs] = get_schema_and_rs(table);
+                auto dc_rf = rs->get_replication_factor(dc);
+                const auto& racks = topology.get_datacenter_rack_nodes().at(dc);
+                auto racks_can_satisfy_their_rf = [&] {
+                    auto rack_rf = std::max(size_t(1), dc_rf / racks.size());
```

Comment:
If you want the maximum number of replicas in a rack, this should just align up the division, like `(dc_rf + (racks.size() - 1)) / racks.size()`, no? Without `std::max`

Reply:
actually, with the conditions below: `racks.size() > 1 && (dc_rf %  racks.size())`, `dc_rf` is a multiple of `racks.size()` so no need to align anything.  We can probably assume `dc_rf > 0`, or check it (it should be an internal error, no, if there are any replicas in the dc and its dc_rf==0).
Anyhow, `dc_rf / racks.size()` should have no remainder and should be greater than or equal to 1, so the `std::max` is not required.

Reply:
Sure.